contracts/Vault.vy

```diff
@@ -1382,3 +1397,52 @@ def sweep(_token: address, _value: uint256 = MAX_UINT256):
     if _value == MAX_UINT256:
         value = ERC20(_token).balanceOf(self)
     self.erc20_safe_transfer(_token, self.governance, value)
+
+
+@view
+@internal
+def permit_digest(owner: address, spender: address, amount: uint256, nonce: uint256, expiry: uint256) -> bytes32:
+    return keccak256(
+        concat(
+            b'\x19\x01',
+            self.DOMAIN_SEPARATOR,
+            keccak256(
+                concat(
+                    PERMIT_TYPE_HASH,
+                    convert(owner, bytes32),
+                    convert(spender, bytes32),
+                    convert(amount, bytes32),
+                    convert(nonce, bytes32),
+                    convert(expiry, bytes32),
+                )
+            )
+        )
+    )
+
+
+@external
+def permit(owner: address, spender: address, amount: uint256, expiry: uint256, signature: Bytes[65]) -> bool:
+    """
+    @notice
+        Approves spender by owner's signature to expend owner's tokens.
+        See https://eips.ethereum.org/EIPS/eip-2612.
+    
+    @param owner The address which is a source of funds and has signed the Permit.
+    @param spender The address which is allowed to spend the funds.
+    @param amount The amount of tokens to be spent.
+    @param expiry The timestamp after which the Permit is no longer valid.
+    @param signature A valid secp256k1 signature of Permit by owner encoded as r, s, v.
+    """
+    assert owner != ZERO_ADDRESS  # dev: invalid owner
```

Comment:
Not strictly necessary, since it shouldn't be possible to `ecrecover` the empty address as the owner's account

Reply:
We need to know `owner` before recovering because their `nonce` is a part of the signed Permit.